### DISCUSSION TITLE
How do I use TextEditor and TextMate for syntax highlighting?

### DISCUSSION THREAD

**Discussion by SCLD-JBuczkowski** [AvaloniaUI/AvaloniaEdit#391] [Q&A]
Hi:

I'm trying to get AvaloniaEdit.TextEditor working with TextMate so I can support syntax highlighting.
I'm using the example from GitHub

I've installed packages:
	AvaloniaEdit.TextMate
	TextMateSharp.Grammars

I'm following this example:

using AvaloniaEdit;
using AvaloniaEdit.TextMate;
using TextMateSharp;
using TextMateSharp.Grammars;

var _textEditor = this.FindControl<TextEditor>("Editor");
var  _registryOptions = new RegistryOptions(ThemeName.DarkPlus);
var _textMateInstallation = _textEditor.InstallTextMate(_registryOptions);

_textMateInstallation.SetGrammar(_registryOptions.GetScopeByLanguageId(_registryOptions.GetLanguageByExtension(".cs").Id));

I'm getting the following compile error in VS:

The type 'IRegistryOptions' is defined in an assembly that is not referenced. 
You must add a reference to assembly 'TextMateSharp, Version=1.0.29.0, Culture=neutral, PublicKeyToken=null'.	

I installed the TextMate package, but still get the same error.
What am I missing here?

**Comment by danipen** [comment]
AvaloniaEdit uses `<TextMateSharpVersion>1.0.56</TextMateSharpVersion>`.

Are you referencing an specific version of `TextmateSharp`? Clean obj first just in case...

**Comment by SCLD-JBuczkowski** [comment]
OK I was trying to use 1.0.56 (the latest release) and I was getting the compile errors.
I tried dropping back to 1.0.29. It compiles and runs without crashing/exceptions.
However, it neither changes the theme nor appears to set the grammar correctly.
I see no syntax highlighting.

Has something changed in the latest releases which makes the existing example obsolete?

**Comment by danipen** [comment]
I don't think so... Does the demo app works for you?
 https://github.com/AvaloniaUI/AvaloniaEdit/tree/master/src/AvaloniaEdit.Demo

**Reply by SCLD-JBuczkowski** [reply]
I was basing my work on the demo in the original ReadMe example.

```
//First of all you need to have a reference for your TextEditor for it to be used inside AvaloniaEdit.TextMate project.
var _textEditor = this.FindControl<TextEditor>("Editor");

var  _registryOptions = new RegistryOptions(ThemeName.DarkPlus);
var _textMateInstallation = _textEditor.InstallTextMate(_registryOptions);

_textMateInstallation.SetGrammar(_registryOptions.GetScopeByLanguageId(_registryOptions.GetLanguageByExtension(".cs").Id));
```
I could not get that to build in VS.

I saw a comment in the posts that NuGet packages were published.
https://github.com/AvaloniaUI/AvaloniaEdit/releases/tag/0.10.12

I was trying to use those packages and I still get the same errors.

I downloaded and pulled the latest and the demo works.

I think that this has something to do with the packages/versions I'm using and can't quite figure it out.

How should I proceed here?

Thx

**Reply by danipen** [reply]
Probably packages are based in old TextMate versions. We probably should generate new nugets. CC/ @Takoooooo 

I recommend using latest sources instead nugets.

**Reply by SCLD-JBuczkowski** [reply]
Cool Beans!

So I should just download, include and build sources locally?

If I grab published ones, in theory when published I should be able to drop and go?
Will I still need TextMateSharp and TextMateSharp.Grammars or just drop and used your published ones?

**Reply by danipen** [reply]
You only need to add [AvaloniaEdit](https://github.com/AvaloniaUI/AvaloniaEdit/tree/master/src/AvaloniaEdit) and [AvaloniaEdit.TextMate](https://github.com/AvaloniaUI/AvaloniaEdit/tree/master/src/AvaloniaEdit.TextMate) (from this repo) to your sources. Then TextMate sources should be referenced automatically using nuget.

**Comment by SCLD-JBuczkowski** [comment]
So I do/don't need to include TextMate directly in my solution?

**Reply by danipen** [reply]
You don't need to include TextmateSharp... But Avalonia.TextMate should be included.

**Comment by danipen** [comment]
Same as the demo project ... https://github.com/AvaloniaUI/AvaloniaEdit/tree/master/src/AvaloniaEdit.Demo

**Reply by SCLD-JBuczkowski** [reply]
It looks like using the locally built versions of the libraries will do the trick.
The only issue is that we are stuck using AvaloniaUI .10.18 due to not being able to fully test moving to 11.x before our release date.
As a result, I'm getting the following error:

Error	CS1705	Assembly 'AvaloniaEdit.TextMate' with identity 'AvaloniaEdit.TextMate, Version=11.0.5.0, Culture=neutral, PublicKeyToken=null' uses 'Avalonia.Base, Version=11.0.0.0, Culture=neutral, PublicKeyToken=c8d484a7012f9a8b' which has a higher version than referenced assembly 'Avalonia.Base' with identity 'Avalonia.Base, Version=0.10.18.0, Culture=neutral, PublicKeyToken=c8d484a7012f9a8b'	

I've tried rebuilding Avalonia.TextMate locally with 0.10.18 and I'm getting compile errors.

Is there any way to get around this versioning collision issue or will I have to wait until we move our app to 11.x?

**Reply by danipen** [reply]
Uh I guess the issue is that latest AvaloniaEdit targets Avalonia 11.x. 

If you are targeting Avalonia 10.x you should go back in history and find the latest commit of AvaloniaEdit targeting 10.x and use that.

**Comment by danipen** [comment]
@SCLD-JBuczkowski https://github.com/AvaloniaUI/AvaloniaEdit/pull/394
